pkg/skaffold/render/renderer/renderer.go

```diff
@@ -92,11 +119,36 @@ func (r *SkaffoldRenderer) Render(ctx context.Context, out io.Writer, builds []g
 	manifests.SetLabels(r.labels)
 
 	// cache the dry manifests to the temp directory. manifests.yaml will be truncated if already exists.
-	dryConfigPath := filepath.Join(r.workingDir, DefaultHydrationDir, dryFileName)
+	dryConfigPath := filepath.Join(r.hydrationDir, dryFileName)
 	if err := manifest.Write(manifests.String(), dryConfigPath, out); err != nil {
 		return err
 	}
 
-	// TODO: mutate and validate
+	// Read the existing Kptfile content. Kptfile is guaranteed to be exist in prepareHydrationDir.
+	kptFilePath := filepath.Join(r.hydrationDir, kptfile.KptFileName)
+	file, err := os.Open(kptFilePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	kfConfig := &kptfile.KptFile{}
+	if err := yaml.NewDecoder(file).Decode(&kfConfig); err != nil {
+		return fmt.Errorf("unable to parse %v\n, please check if the kptfile is updated to new apiVersion > v1alpha2",
```

Comment:
thanks for creating an actionable error. Can you use sErrors.New() and add error code as well as suggestion code to this error. 
Docs here
https://github.com/GoogleContainerTools/skaffold/blob/master/DEVELOPMENT.md#adding-actionable-error-messages-to-code

The advantage of this this is, IDEs can consume the suggestion code and decide to show an actionable error message to the user as is or relevant one.

Reply:
oh, we do not want to add the err. If the Decode fails, it basically means the yaml is invalid, and no more details will be given from the Decode return error. That's why we add specific user actions to help users triage.

Reply:
we still need an error code with an actionable error message `KPT_INVALID_YAML` or `KPT_USER_ERROR` so we can track these in skaffold and get metrics on it. 
Right, if a skaffold dev fails, the error code logged via metrics pipeline  will be `DEPLOY_UNKNOWN`. We are trying to reduce these overall.

Reply:
I see. I added some TODO comments (here and other errors in ./render). Will add the Suggestion and error code in the following CL.

Reply:
@tejal29 Here's the proto PR https://github.com/GoogleContainerTools/skaffold/pull/5953